### PR TITLE
C# Improvements

### PR DESCRIPTION
**Pull request by eisber**
@lhoang29  review :)

allow models to be saved
added namespace/feature enumeration support from C# side
allow ExampleBuilder to build against example pool vs vw instance
exposing scalars (float[]) predictions
support schema specification for ADF types
support labels to be discovered by serializer
support multiclass label
allow for graceful handling of uninitialized thread pool
basic support for config files (ADF missing)
added documentation
improved nuget/msbuild target wrt zlib.dll. properly cascades zlib dependency
allow retargeting of examples from creation instance to learn/prediction instance. useful to split serialization/learning into multiple threads
added threadsafe example pool

### CODE REVIEW

**Comment by JohnLangford** [comment]
@lhoang29 ping

**Comment by JohnLangford** [comment]
Markus, it would also help to have a bit more description of what's going on here.

**Comment by JohnLangford** [comment]
(Is this a big collection of little changes?)

**Comment by eisber** [comment]
I put the following comment in. Are you looking for more?

@lhoang29<https://github.com/lhoang29> review :)
allow models to be saved
added namespace/feature enumeration support from C# side
allow ExampleBuilder to build against example pool vs vw instance
exposing scalars (float[]) predictions
support schema specification for ADF types
support labels to be discovered by serializer
support multiclass label
allow for graceful handling of uninitialized thread pool
basic support for config files (ADF missing)
added documentation
improved nuget/msbuild target wrt zlib.dll. properly cascades zlib dependency
allow retargeting of examples from creation instance to learn/prediction instance. useful to split serialization/learning into multiple threads
added threadsafe example pool

**Comment by eisber** [comment]
Yes.

File: vowpalwabbit/mwt.cc [hunk 1]
```diff
@@ -154,30 +148,68 @@ namespace MWT {
...
+
+    stringstream msg;
+
+    // total
+    msg << "total: " << c.total;
+    bin_text_read_write_fixed_validated(model_file, (char*)&c.total, sizeof(c.total), "", read, msg, text);
+
+    // policies
+    size_t policies_size = c.policies.size();
+    bin_text_read_write_fixed_validated(model_file, (char*)&policies_size, sizeof(policies_size), "", read, msg, text);
```
Comment (JohnLangford) [inline vowpalwabbit/mwt.cc]: You need to update msg here, right?

File: vowpalwabbit/mwt.cc [hunk 2]
```diff
@@ -226,6 +259,7 @@ base_learner* mwt_setup(vw& all)
   else
     l = &init_learner(&c, setup_base(all), predict_or_learn<false, false, true>, predict_or_learn<false, false, false>, 1);
 
+  l->set_save_load(save_load);
   l->set_finish_example(finish_example);
```
Comment (JohnLangford) [inline vowpalwabbit/mwt.cc]: Huh.  Set_finish_example _is_ set here.

Comment (JohnLangford) [inline vowpalwabbit/mwt.cc]: Ignore---was confusing with cb_explore

**Comment by JohnLangford** [comment]
@lhoang29 have you had a chance to review?

I'll reiterate: large patches are bad.  They may incorporation much more difficult because there are many more things that have to be done right.

File: vowpalwabbit/mwt.cc (hunk 1 above)
Comment (eisber) [inline vowpalwabbit/mwt.cc]: No, bin_text_read_write_fixed_validated sets msg to “”

**Comment by eisber** [comment]
Sorry…

**Comment by lhoang29** [comment]
oops just saw the email, don't know why Github didn't notify me on this one. I'll take a look tomorrow

File: cs/common/Reflection/ReflectionHelper.cs [hunk 3]
```diff
@@ -147,7 +147,7 @@ select new
                 method = method.MakeGenericMethod(actualTypes);
                 //Debug.WriteLine("\t specializing: " + method);
             }
-            //Debug.WriteLine("Method: {0} for {1} {2}", method, name, string.Join(",", parameterTypes.Select(t => t.ToString())));
+            // Debug.WriteLine("Method: {0} for {1} {2}", method, name, string.Join(",", parameterTypes.Select(t => t.ToString())));
```
Comment (lhoang29) [inline cs/common/Reflection/ReflectionHelper.cs]: remove?

File: cs/cs/Serializer/ConfigInspector.cs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                    (?<attrname>\w+) \s* = \s* (?<attrvalue>[^,)\s]+)
+                    (?:\s*,\s*   # other attributes
+                        (?<attrname>\w+) \s* = \s* (?<attrvalue>[^,)\s]+) \s*
+                    )*
+                \))?
+                \s*$
+                ",
+            RegexOptions.IgnorePatternWhitespace | RegexOptions.Compiled);
+
+        public static Schema CreateSchema(Type type, string config, Action<string> warnings = null)
```
Comment (lhoang29) [inline cs/cs/Serializer/ConfigInspector.cs]: can't seem to make out what this method is for?

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs [hunk 5]
```diff
@@ -102,6 +102,47 @@ public void Parse(string json, ILabel label = null, int? index = null)
...
+            // handle the case when the reader is already positioned at JsonToken.StartObject
+            if (reader.TokenType == JsonToken.None && !reader.Read())
+                throw new VowpalWabbitJsonException(reader.Path, "Expected non-empty JSON");
+
+            if (reader.TokenType != JsonToken.StartObject)
+                throw new VowpalWabbitJsonException(reader.Path, "Expected start object");
+
+            while (reader.Read())
+            {
+                if (!(reader.TokenType == JsonToken.PropertyName && (string)reader.Value == "_multi"))
```
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: don't you have this string defined somewhere?

File: cs/common/Reflection/ReflectionHelper.cs (hunk 3 above)
Comment (eisber) [inline cs/common/Reflection/ReflectionHelper.cs]: This is really helpful from time to time to debug things. Maybe I should allow for tracing? Do you know about standard ways of configuring tracing for C# libs?

File: cs/cs/Serializer/ConfigInspector.cs (hunk 4 above)
Comment (eisber) [inline cs/cs/Serializer/ConfigInspector.cs]: That’s for parsing the dynamic feature config file that Oswaldo asked for.

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs (hunk 5 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: Hmm. Actually that’s the string I wanted to make configurable. What do you think? Add as settings?

File: cs/cs/VowpalWabbitMultiworldTesting.cs [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+                    return this.data.Skip(1);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Must public for the serializer to work with.
+        /// </summary>
+        [EditorBrowsableAttribute(EditorBrowsableState.Never)]
+        public sealed class LearnedVsConstantPolicy
```
Comment (lhoang29) [inline cs/cs/VowpalWabbitMultiworldTesting.cs]: hmm, not sure what I think about this. It seems a bit awkward to have a class for just the learned and constant-action policies. Would it be better to do this generally so you just have a single class with a feature vector. The caller can then decide what this vector means? (for example, in this case the first one is learned and rest is constant-action)

File: cs/unittest/TestConfigInspector.cs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    {
+        [TestMethod]
+        public void TestConfigParsing()
+        {
+            var str = @"
+f1
+ns1.f2(FeatureGroup = 'a')
+ns1.f3(Namespace = ""a"",Enumerize=true, FeatureGroup = 'g'  )                
+ns1.f4(Enumerize=true , FeatureGroup ='f', Order=3,AddAnchor=true)   
+ns1.ns2.f5
```
Comment (lhoang29) [inline cs/unittest/TestConfigInspector.cs]: maybe this answers my question above. This is for ingesting the dynamic context structure?

File: cs/common/Reflection/ReflectionHelper.cs (hunk 3 above)
Comment (lhoang29) [inline cs/common/Reflection/ReflectionHelper.cs]: doesn't Tracing write to VW output window by default in debug mode?

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs (hunk 5 above)
Comment (lhoang29) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: yeah seems like a good idea

File: cs/common/Reflection/ReflectionHelper.cs (hunk 3 above)
Comment (eisber) [inline cs/common/Reflection/ReflectionHelper.cs]: I'm more asking about a more general strategy for tracing in the library.

File: cs/unittest/TestConfigInspector.cs (hunk 7 above)
Comment (eisber) [inline cs/unittest/TestConfigInspector.cs]: yes.

File: cs/cs/Serializer/VowpalWabbitJsonSerializer.cs (hunk 5 above)
Comment (eisber) [inline cs/cs/Serializer/VowpalWabbitJsonSerializer.cs]: implemented

File: cs/cs/VowpalWabbitMultiworldTesting.cs (hunk 6 above)
Comment (eisber) [inline cs/cs/VowpalWabbitMultiworldTesting.cs]: agree

**Pull request merged**